transform_assumption.ml

```diff
@@ -1,37 +1,239 @@
+ * A checkpoint is a well-formed osr point and a matching label.
+ * For example:
+ *   checkpoint_1:
+ *     osr [] (func, version, checkpoint_1) [mut x = &x]
+ * Initially the checkpoint is self referential (ie. its target is
+ * the label right above. Insert assumption makes use of this property
+ * to create a new version where the checkpoints are in sync.
+ * This has to run before any optimistic optimizations. *)
+let insert_checkpoints (func:afunction) =
+  (* Can only insert checkpoints into an unoptimized function *)
+  if List.length func.body <> 1 then None else
 
-let insert_branch_pruning_assumption (func : afunction) : version option =
-  let version = Instr.active_version func in
+  let version = List.hd func.body in
   let instrs = version.instrs in
   let inp = Analysis.as_analysis_input func version in
   let scope = Scope.infer inp in
   let live = Analysis.live inp in
+
   let transform pc =
-    match scope.(pc) with
-    | DeadScope -> assert(false)
-    | Scope scope ->
-      begin match[@warning "-4"] instrs.(pc) with
-      | Branch (exp, l1, l2) ->
+    let create_checkpoint pc =
+      match scope.(pc) with
+      | DeadScope -> assert(false)
+      | Scope scope ->
         let osr = List.map (function
-            | (Const_var, x) ->
-              Osr_const (x, (Simple (Var x)))
-            | (Mut_var, x) ->
-              if List.mem x (live pc) then
-                Osr_mut_ref (x,  x)
-              else
-                Osr_mut_undef x)
-            (ModedVarSet.elements scope)
-        in
-        Insert [Osr (exp, func.name, version.label, l1, osr); Goto l2]
+          | (Const_var, x) -> Osr_const (x, (Simple (Var x)))
+          | (Mut_var, x)   ->
+            if List.mem x (live (pc-1)) then
+              Osr_mut_ref (x, x)
+            else
+              Osr_mut_undef x) (ModedVarSet.elements scope) in
+        let target = { func=func.name; version=version.label; label=checkpoint_label pc } in
+        Insert [Label (checkpoint_label pc); Osr {cond=[]; target; map=osr};]
+    in
+    if pc = 0 then Unchanged else
+    match[@warning "-4"] instrs.(pc) with
+    | Stop _ | Return _ | Label _ | Comment _ -> Unchanged
+    | Osr _ -> assert false
+    | _ -> create_checkpoint pc
+  in
+  let baseline = change_instrs transform inp in
+  let (|?) opt def = match opt with Some v -> v | None -> def in
+  Some { func with
+         body = [
+           { label = version.label;
+             instrs = (|?) baseline instrs;
+             annotations = None } ] }
+
+let remove_empty_osr ({instrs} as inp) : instructions option =
+  let transform pc =
+    match[@warning "-4"] instrs.(pc) with
+    | Osr {cond} when cond = [] -> Remove 1
+    | _ -> Unchanged
+  in
+  change_instrs transform inp
+
+(* Removes all checkpoint labels. Can be applied as a final cleanup
+ * when there are no osr-in points. *)
+let remove_checkpoint_labels ({instrs} as inp) : instructions option =
```

Comment:
agreed. I create an issue for it (I'd rather not completely disable it because it makes the output of the optimizer super annoying to read...)